### PR TITLE
Vine: Make conservative disk estimates when the amount of disk is not given to worker

### PR DESCRIPTION
**Pull request by tphung3**
## Proposed changes

When a task with no resources specified lands on a worker, it is assigned all resources of the worker (see `normalize_resources`). However, later on when the worker checks whether the task can fit the current resources, some resources have been taken away by other tasks, so the current task is too big for the worker and thus it is forsaken.

Example: an unspecified task arrives and is assigned 4 GBs of disk space. It depends on a poncho package of size 300 MBs. Once the poncho package is unpacked, the worker later finds out that its disk capacity is only 4 GBs - 300 MBs, so it forsakens the unspecified task as it thinks it doesn't have enough disk space while the unspecified task should run anyway.

Note that this problem doesn't happen with tasks that are clean, i.e., once their executions are done, their workspaces are wiped and resources are returned. Tasks that have side-effects like minitasks use up some disk space so they interact weirdly with regular tasks with unspecified resources.

This problem goes undetected until now probably because the poncho package is too small (<1 MB) or the task passes the resource check before the worker updates its disk counting.

A check whether library name exists is added to prevent seg fault.

## Post-change actions

Put an 'x' in the boxes that describe post-change actions that you have done.
The more 'x' ticked, the faster your changes are accepted by maintainers.

- [x] `make test`       Run local tests prior to pushing.
- [x] `make format`     Format source code to comply with lint policies. Note that some lint errors can only be resolved manually (e.g., Python)
- [x] `make lint`       Run lint on source code prior to pushing.
- [x] Type Labels       Select github labels for the type of this change: bug, enhancement, etc.
- [x] Product Labels    Select github labels for the product affected: TaskVine, Makeflow, etc.
- [x] PR RTM            Mark your PR as ready to merge.

## Additional comments
This section is dedicated to changes that are ambitious or complex and require substantial discussions. Feel free to start the ball rolling.

### CODE REVIEW

**Comment by btovar** [comment]
@tphung3 the poncho packages I use are around 300MB, and expand to 1.4 GB.  The forsaken result occurs when the worker machine has less resources than originally promised. In your tests, are you specifying the `--disk` parameter?  Workers are pretty much unusable right now without specifying the `--disk` parameter.  We do not observe this issue with the submit scripts, or with the factory because they set `--disk` parameter to some value less than the whole machine.

**Comment by btovar** [comment]
I'd suggest not removing the check for disk space. A better solution is to set the default value the disk to some percentage less than the whole machine. (Say, report 75% of the disk if `--disk` is not specified, and add an option to change this percentage.).

**Comment by tphung3** [comment]
> @tphung3 the poncho packages I use are around 300MB, and expand to 1.4 GB. The forsaken result occurs when the worker machine has less resources than originally promised. In your tests, are you specifying the `--disk` parameter? Workers are pretty much unusable right now without specifying the `--disk` parameter. We do not observe this issue with the submit scripts, or with the factory because they set `--disk` parameter to some value less than the whole machine.

I did not set a disk value for workers.

**Comment by dthain** [comment]
Right, I think our mistake is not in the checking of the resources, but in the assigning 100% of disk space to a task.  It's ignoring the disk space assigned to caching.

In general, we need to take a careful look at how disk space allocation is performed, because we are being careful about counting some aspects, and not so careful in others.

**Comment by tphung3** [comment]
The 90% safeguard to disk estimates now remove the task forsaking behavior.

File: taskvine/src/worker/vine_worker.c [hunk 1]
```diff
@@ -320,6 +326,11 @@ static void measure_worker_resources()
 
 	if (manual_disk_option > 0) {
 		r->disk.total = MIN(r->disk.total, manual_disk_option);
+	} else {
+		/* Set the reporting disk to a fraction of the measured disk to avoid
+		 * unnecessarily forsaking tasks with unspecified resources. */
+		manual_disk_option = floor(r->disk.total * disk_percent / 100);
```
Comment (btovar) [inline taskvine/src/worker/vine_worker.c:332]: Protect for values outise [0,100]:

r->disk.total = ceil(r->disk.total * MIN(100, MAX(disk_percent, 0)));

and let's leave manual_disk_option unset to mark that the disk was not specified from the command line.

File: taskvine/src/worker/vine_worker.c [hunk 2]
```diff
@@ -2088,6 +2097,10 @@ static void show_help(const char *cmd)
 	printf(" %-30s Manually set the amount of disk (in MB) reported by this worker.\n", "--disk=<mb>");
 	printf(" %-30s If not given, or less than 1, then try to detect disk space available.\n", "");
 
+	printf(" %-30s Set the conservative disk reporting percent when amount of disk is unspecified.\n",
```
Comment (btovar) [inline taskvine/src/worker/vine_worker.c:2100]: when --disk is unspecified.

File: taskvine/src/worker/vine_worker.c [hunk 3]
```diff
@@ -1170,7 +1179,7 @@ struct vine_process *find_library_for_function(const char *library_name)
 
 	ITABLE_ITERATE(procs_running, task_id, p)
 	{
-		if (!strcmp(p->task->provides_library, library_name)) {
+		if (p->task->provides_library && !strcmp(p->task->provides_library, library_name)) {
```
Comment (btovar) [inline taskvine/src/worker/vine_worker.c:1182]: should this be included in another pr? It looks unrelated.

**Review by btovar — CHANGES_REQUESTED** [review]

**Review by btovar — APPROVED** [review]

**Pull request merged**